Feature: saved launchtime configurations

AlanCoding opened the pull request:
I still hope to do manual testing with tower-cli (corresponding PR already exists in that repo), but all of my TODO items for the feature are finished, and I've been doing some manual testing.

Connect #169 (the main one)

Connect #497 
Connect #555 
Connect #478

AlanCoding committed:
Feature: saved launchtime configurations

Consolidate prompts accept/reject logic in unified models
Break out accept/reject logic for variables
Surface new promptable fields on WFJT nodes, schedules

Make schedules and workflows accurately reject variables
  that are not allowed by the prompting
  rules or the survey rules on the template

Validate against unallowed extra_data in system job schedules
Prevent schedule or WFJT node POST/PATCH with unprompted data
Move system job days validation to new mechanism
Add new psuedo-field for WFJT node credential
Add validation for node related credentials
Add related config model to unified job
Use JobLaunchConfig model for launch RBAC check

Support credential overwrite behavior with multi-creds
  change modern manual launch to use merge behavior
Refactor JobLaunchSerializer, self.instance=None
Modularize job launch view to create "modern" data
Auto-create config object with every job
Add create schedule endpoint for jobs

AlanCoding commented:
Here's what some of the re-launch schedules look like in the schedule list

![screen shot 2017-11-29 at 1 56 21 pm](https://user-images.githubusercontent.com/1385596/33393433-32368230-d50d-11e7-8169-5be0e23ab63c.png)

I can make that say whatever people want, but I wasn't really into the idea of using a timestamp like copy. I like how clean this is.

The `- number` is when you create more than 1 schedule from the same job (not saying this is advisable).

AlanCoding committed:
Tweak validation to allow multiple vault credentials

support providing vault passwords based on id
include needed passwords in launch serializer defaults

AlanCoding commented:
TODO:

From conversations with @mabashian, we're going to be best off to move to a model where we show the needed password inside of the credential summary inside of `defaults`. This might be mandatory in order for the UI to do housekeeping about which credentials will be present on the job ultimately.

And the migration has already been rebased on top of multi-vault. Since that TODO intersects with multi-vault already, I might as well go ahead and make some to test with, and the pass-though of credentials from the POST data might be practically automatic.

AlanCoding commented:
I tried to prettify the credential `unique_hash` kind of stuff for display. Here's what the error will look like now:

```json
{
    "msg": "Cannot assign multiple vault (id=alan) credentials."
}
```

This also gives us room to make the label more verbose if requested in the future.

AlanCoding commented on awx/main/models/credential.py:
```diff
@@ -326,9 +326,14 @@ def needs_vault_password(self):
+            if self.inputs.get('vault_id'):
+                needed.append('vault_password.{}'.format(self.inputs.get('vault_id')))
+            else:
+                needed.append('vault_password')
```
@ryanpetrello this is something you specifically referred to doing (I wanted to do this at the same time as connecting needed creds to the cred list), and I think it takes care of it, because this is the same string format that tasks.py is expecting. Haven't tested _fully_, but I don't have a reason to say it won't work anymore.

ryanpetrello replied:
👏 I'll test this all post-merge, but this _should be_ all we need for multi vault to be called done.  Thanks for this!

AlanCoding committed:
combine launch config and multi-cred migrations

ryanpetrello commented on docs/prompting.md:
```diff
@@ -0,0 +1,253 @@
+Prompting enablement for several types of credentials is controlled by a single
+field. On launch, multiple types of credentials can be provided in their respective fields
+inside of `credential`, `vault_credential`, and `extra_credentials`. Providing
+a credential that requirements password input from the user on launch is
```
s/requirements/requires

ryanpetrello commented on awx/main/models/credential.py:
```diff
@@ -369,6 +374,31 @@ def display_inputs(self):
                 field_val[k] = '$encrypted$'
         return field_val
 
+    def unique_hash(self, display=False):
```
Does anything call `unique_hash(display=True)`?  If not, do we need this extra logic? Seems like we could simplify this method quite a bit:

```python
def unique_hash(self):
    uniq = str(self.credential_type_id)
    if self.kind == 'vault' and self.inputs.get('vault_id', None):
        return '_'.join((uniq, self.inputs['vault_id']))
    return uniq
```

AlanCoding replied:
Yeah, I started doing that for the view (when returning error messages). I should document what the 2 look like, if we keep the dichotomy

display format: `vault (id=alan)`
other format: `3_alan` (assuming 3 is the pk of the vault type)

I'd vote that we should keep the display format if we're going to use long names like literally "Amazon Web Services".

AlanCoding replied:
Okay, so integration tests have told me what the expectation is (`test_confirm_only_single_managed_by_tower_extra_credential_allowed`). They are expecting the error message:

```python
{'error': 'Cannot assign multiple VMware vCenter credentials.'}
```

This is pre-existing in 3.2, so easiest to just keep it the same.

ryanpetrello commented on awx/main/models/unified_jobs.py:
```diff
@@ -357,18 +366,48 @@ def create_unified_job(self, **kwargs):
+            cred_dict = Credential.unique_dict(self.credentials.all())
+            prompted_dict = Credential.unique_dict(kwargs['credentials'])
+            # combine prompted credentials with JT
+            cred_dict.update(prompted_dict)
```
`dict.update` is a smart, performant way to solve the merge-by-"kind" problem _including_ the vault special case, good work 👍

ryanpetrello commented on awx/main/models/unified_jobs.py:
```diff
@@ -357,18 +366,48 @@ def create_unified_job(self, **kwargs):
+
         return unified_job
 
+    @cached_subclassproperty
```
Is calculating this really so expensive that we need to cache it?  This seems like needless complexity.

AlanCoding replied:
I agree caching is unimportant here. But the problem is that I still prefer something that can act as a class property, and it looks like it has to be custom. I don't know how to get that without defining a descriptor. I could change it to `subclassproperty` (and considered doing this), but that still leaves almost all of the added complexity.

ryanpetrello replied:
It _usually_ pains me to recommend a Python metaclass, but this actually a perfect example of where you should use one.  You could accomplish something similar by using a metaclass instead of your custom class decorator approach in combination w/ the property you described at https://github.com/ansible/awx/pull/740#discussion_r155324067:

```diff
diff --git a/awx/main/models/unified_jobs.py b/awx/main/models/unified_jobs.py
index d19af29..48033a6 100644
--- a/awx/main/models/unified_jobs.py
+++ b/awx/main/models/unified_jobs.py
@@ -36,8 +36,7 @@ from awx.main.models.mixins import ResourceMixin, TaskManagerUnifiedJobMixin
 from awx.main.utils import (
     decrypt_field, _inventory_updates,
     copy_model_by_class, copy_m2m_relationships,
-    get_type_for_model, parse_yaml_or_json,
-    cached_subclassproperty
+    get_type_for_model, parse_yaml_or_json
 )
 from awx.main.redact import UriCleaner, REPLACE_STR
 from awx.main.consumers import emit_channel_notification
@@ -51,11 +50,29 @@ CAN_CANCEL = ('new', 'pending', 'waiting', 'running')
 ACTIVE_STATES = CAN_CANCEL
 
 
+class UnifiedJobTemplateMeta(type(PolymorphicModel), type(CommonModelNameNotUnique)):
+
+    @property
+    def ask_mapping(cls):
+        return {
+            (field.allows_field, field.name) for field in cls._meta.fields
+            if isinstance(field, AskForField)
+         }
+
+
 class UnifiedJobTemplate(PolymorphicModel, CommonModelNameNotUnique, NotificationFieldsModel):
     '''
     Concrete base class for unified job templates.
     '''
 
+    __metaclass__ = UnifiedJobTemplateMeta
+
     # status inherits from related jobs. Thus, status must be able to be set to any status that a job status is settable to.
     JOB_STATUS_CHOICES = [
         ('new', _('New')),                  # Job has been created, but not started.
@@ -395,19 +412,6 @@ class UnifiedJobTemplate(PolymorphicModel, CommonModelNameNotUnique, Notificatio
 
         return unified_job
 
-    @cached_subclassproperty
-    def ask_mapping(cls):
-        mapping = {}
-        for field in cls._meta.fields:
-            if not isinstance(field, AskForField):
-                continue
-            if field.allows_field == '__default__':
-                allows_field = field.name[len('ask_'):-len('_on_launch')]
-            else:
-                allows_field = field.allows_field
-            mapping[allows_field] = field.name
-        return mapping
-
     @classmethod
     def _get_unified_jt_copy_names(cls):
         return cls._get_unified_job_field_names()
diff --git a/awx/main/utils/common.py b/awx/main/utils/common.py
index cdb84b2..e1b414e 100644
--- a/awx/main/utils/common.py
+++ b/awx/main/utils/common.py
@@ -47,7 +47,7 @@ __all__ = ['get_object_or_400', 'get_object_or_403', 'camelcase_to_underscore',
            'extract_ansible_vars', 'get_search_fields', 'get_system_task_capacity',
            'wrap_args_with_proot', 'build_proot_temp_dir', 'check_proot_installed', 'model_to_dict',
            'model_instance_diff', 'timestamp_apiformat', 'parse_yaml_or_json', 'RequireDebugTrueOrTest',
-           'has_model_field_prefetched', 'set_environ', 'IllegalArgumentError', 'cached_subclassproperty',]
+           'has_model_field_prefetched', 'set_environ', 'IllegalArgumentError']
 
 
 def get_object_or_400(klass, *args, **kwargs):
@@ -935,17 +935,3 @@ def has_model_field_prefetched(model_obj, field_name):
     # NOTE: Update this function if django internal implementation changes.
     return getattr(getattr(model_obj, field_name, None),
                    'prefetch_cache_name', '') in getattr(model_obj, '_prefetched_objects_cache', {})
-
-
-class cached_subclassproperty(object):
-    '''Caches property in subclasses'''
-
-    def __init__(self, method):
-        self.method = method
-        self.name = method.__name__
-
-    def __get__(self, instance, cls):
-        r = self.method(cls)
-        if self.name not in cls.__dict__:
-            setattr(cls, self.name, r)
-        return r
```

AlanCoding replied:
What if I just made it a `classmethod`, and called it like `JobTemplate.ask_mapping()` everywhere?

ryanpetrello replied:
That's fine, too; it just sounded like you _really_ wanted a `@property`-like interface here.  A _much_ simpler approach is to just make this a `classmethod`.  If you're going to change it to a method though, please change the name to something method-like, such as `get_ask_mapping()` 😄

ryanpetrello commented on awx/main/models/unified_jobs.py:
```diff
@@ -357,18 +366,48 @@ def create_unified_job(self, **kwargs):
+        for field in cls._meta.fields:
+            if not isinstance(field, AskForField):
+                continue
+            if field.allows_field == '__default__':
```
Instead of using a special string here `__default__` and at https://github.com/ansible/awx/pull/740/files#diff-ea6da88b8c0cd3fbf7858ae87933b296R770, why not just use `None`?  That way the `allows_fields=...` argument is how you specify behavior that's *different* from the default.

Instead of building this mapping every time you need it and dealing with caching in this function, why not construct it and store in the `AskForField` object itself e.g.,

Change this code:
https://github.com/ansible/awx/pull/740/files#diff-ea6da88b8c0cd3fbf7858ae87933b296R770

...to do this:

```python
def __init__(self, allows_field=None, **kwargs):
    super(AskForField, self).__init__(**kwargs)
    self.allows_field = allows_field or self.name[len('ask_'):-len('_on_launch')]
```

...and change the `ask_mapping` function to do this:

```python

def ask_mapping(cls):
    return {
        (field.allows_field, field.name) for field in cls._meta.fields
        if isinstance(field, AskForField)
    }
```

AlanCoding replied:
Probably need to add a property on the AskForField field class to complement that, like

```python
@property
def allows_field(self):
    return self._allows_field if self._allows_field is not None else self.name
```

Agree - something with more of this structure would look great.

We can't reference `self.name` on initialization, but it won't be a problem with this approach.

ryanpetrello replied:
Well...

```python
... else self.name[len('ask_'):-len('_on_launch')
```

Yes?

AlanCoding replied:
ugh, yes, that. I only claim to meta-code in github comments.

ryanpetrello commented on awx/main/utils/common.py:
```diff
@@ -934,3 +935,17 @@ def has_model_field_prefetched(model_obj, field_name):
                    'prefetch_cache_name', '') in getattr(model_obj, '_prefetched_objects_cache', {})
+
+
+class cached_subclassproperty(object):
```
Curious if we really need this for what we're using it for given my other comments.

ryanpetrello commented on awx/main/models/jobs.py:
```diff
@@ -350,66 +341,72 @@ def can_start_without_user_input(self, callback_extra_vars=None):
                 not self.passwords_needed_to_start and
                 not variables_needed)
 
-    def _ask_for_vars_dict(self):
```
I really like that we're replacing this with something dynamic, and specifying exceptions *at the place* the field is defined 👍 , it's a lot easier to follow, in my opinion.

ryanpetrello commented on awx/main/models/unified_jobs.py:
```diff
@@ -389,6 +428,37 @@ def copy_unified_jt(self):
+        according to the acceptance rules of the template.
+        '''
+        if errors is None:
+            errors = {}
```
:)

ryanpetrello commented on awx/main/models/jobs.py:
```diff
@@ -793,6 +790,160 @@ def finish_job_fact_cache(self):
+            extra_data = json.loads(self.extra_data)
+            for key, value in self.survey_passwords.items():
+                if key in extra_data:
+                    extra_data[key] = value
```
We'll have to consolidate this w/ the 3.2.2 survey encryption work when it merges back in to awx.

ryanpetrello commented on awx/main/models/jobs.py:
```diff
@@ -793,6 +790,160 @@ def finish_job_fact_cache(self):
+        Only used for workflow nodes to support backward compatibility.
+        '''
+        try:
+            return [cred for cred in self.credentials.all() if cred.credential_type.kind == 'ssh'][0]
```
```python
return list(self.credentials.filter(credential_type__kind='ssh'))[0]
```

ryanpetrello commented on awx/main/models/workflow.py:
```diff
@@ -186,19 +127,29 @@ def create_wfjt_node_copy(self, user, workflow_job_template=None):
+        allowed_creds = []
         for field_name in self._get_workflow_job_field_names():
+            if field_name == 'credentials':
+                Credential = self._meta.get_field('credentials').related_model
```
What's the need to look up the class this way?  Something to make migrations happy?

AlanCoding replied:
Right, I do this in several places. It's because of the import order. If unified_jobs.py imports Credential and credential.py imports UnifiedJob, then you get a startup server error. This was most important for the UnifiedJob <-> Schedule relationships. But here I'm just trying to avoid putting credentials higher up the chain the workflows. Granted, we could probably safely import Credential here, because workflow should be unlikely to be imported by other models (it's kind of the last one down the chain).

ryanpetrello commented on awx/api/serializers.py:
```diff
@@ -2981,104 +3018,160 @@ class Meta:
 
 
-class WorkflowNodeBaseSerializer(BaseSerializer):
-    job_type = serializers.CharField(allow_blank=True, allow_null=True, required=False, default=None)
```
Is it possible to have an old WorkflowNode w/ a `job_type` _not in_ `NEW_JOB_TYPE_CHOICES`?

AlanCoding replied:
do you mean a pre-migration node? The possibility in question is "scan", which isn't valid as-of 3.2, but we may have never removed it with a migration. Valid concern.

ryanpetrello replied:
Yep, nodes that exist prior to 3.2.

ryanpetrello commented on awx/api/serializers.py:
```diff
@@ -2981,104 +3018,160 @@ class Meta:
+
+    def create(self, validated_data):
+        deprecated_fields = {}
+        if 'credential' in validated_data:
```
May want TODOs on some of this backwards-combat stuff signaling that we can drop it at some point when we remove support for this.

ryanpetrello commented on awx/api/serializers.py:
```diff
@@ -3338,114 +3436,86 @@ def get_survey_enabled(self, obj):
                             name=cred.name,
-                            credential_type=cred.credential_type.pk
+                            credential_type=cred.credential_type.pk,
+                            passwords_needed=cred.passwords_needed
```
I like this.

ryanpetrello commented on awx/api/serializers.py:
```diff
@@ -3338,114 +3436,86 @@ def get_survey_enabled(self, obj):
-        for field in obj.resources_needed_to_start:
-            if not (attrs.get(field, False) and obj._ask_for_vars_dict().get(field, False)):
-                errors[field] = _("Job Template '%s' is missing or undefined.") % field
+        template._is_manual_launch = True  # TODO: hopefully remove this
```
This still a hopefully :)?

ryanpetrello commented on awx/api/serializers.py:
```diff
@@ -3338,114 +3436,86 @@ def get_survey_enabled(self, obj):
+        passwords = attrs.get('credential_passwords', {})  # get from original attrs
+        passwords_lacking = []
+        for cred in launch_credentials:
+            if cred.passwords_needed:
```
Can we lose this `if`?  Seems like `passwords_needed` will fall back to just being an empty list and the zero-item loop will have the same effect.

AlanCoding committed:
Make ask_mapping a simple class property

from PR feedback of saved launchtime configurations

AlanCoding commented:
Things that I put a 👍  on should be addressed. Some notes...

What are the drivers for the size of this diff?

- Moving WFJT node to a common launch config base class
  - serializers totally refactored for child models
  - access classes totally refactored for child models
  - model changes + concrete job launch config
- moving roughly 70% of launch validation logic into unified model (prompts processing)
  - also break out logic for variables, this is to reuse for WFJTs, which work differently
- psuedo-field for WFJT credential
- docs
- migration including AskForField (which is a no-op)


PR feedback that was included:
 - workflows import Credential
 - small docs tweak
 - don't bother checking `if cred.passwords_needed`
 - add TODOs for `credential` field removal
 - modify TODOs for hacky parameters set on JT object for prompts special-cases
   - this is because firstly, use of `ignored_fields`, which are error-less on manual launch (not a battle I want to fight here), and because manual launch allows credentials with password prompts, whereas saved launch configurations can not take them
 - remove cached_subclassproperty
 - add migration to remove the scan job_type prompt in nodes, @chrismeyersfsu this is a minor part of the fact searching changes in 3.2, but gets rolled into the data migrations in this feature now

~~TODO: cut out `credential_passwords` not in the required list~~ Never mind, I had already done that. Ignore some of the latest commit noise.

Credential passwords has weird history. Our job launch calls `job.signal_start(**request.data)`, more-or-less. This `request.data` becomes the starting `kwargs` inside of the `run` gauntlet. It feels like that _should_ be very important to preserve... but our own code contradicts this. If the miscellaneous prompts were necessary to have in this data, then relaunch wouldn't have worked.

Because of that, I'm trying to formalize the notion of `signal_start` takes passwords, and `create_unified_job` takes field overrides (of which "prompts" are a subset). Passing things other than passwords could pose a security risk. And of course, in #746, I am trying to kill the permanence of the passwords blob entirely. This is my answer to "why" we can't save credential passwords in the launch configs - because we don't save credential passwords, because it is a bad idea.

Here's an example of `start_args` for a job currently, we don't need (and shouldn't store) this data:

```
In [8]: decrypt_field(j, 'start_args')
Out[8]: '{"survey_passwords": {"a": "$encrypted$", "c": "$encrypted$", "b": "$encrypted$", "d": "$encrypted$"}, "extra_vars": {"c": "asdf", "d": "asdf"}, "launch_type": "workflow"}'
```

I'll squash these commits before merge, FYI.

AlanCoding committed:
Bug fixes from integration ran on launchtime branch

Make error message for muti-vault validation more
consistent with historical message

AlanCoding commented:
I ran integration against this branch locally. I just did the "workflows" tests for now. Here is some output:

https://gist.github.com/AlanCoding/868c36dbf022df5bba114de12788fa3f

This clearly has found some bugs, but I can't really get much value out of it other than the responses and the logs, because it deletes everything that it creates. After I figure out how to hunt these down, I should have more fixes coming for the bugs that are obvious from this output.

 -  assert 'This field cannot be blank.' in "{u'unified_job_template': [u'Node needs to have a template attached.']}"
    - solved
 - BadRequest: {u'all': [u'Fields [] are not allowed on launch.']}
    - should be solved, need to double-check
 - several timeout failures
    - no idea
 - AssertionError: Workflow job 229 unsuccessful
    - no idea
 - Failed: DID NOT RAISE <class 'towerkit.exceptions.BadRequest
    - looks like something solved in 3.2.2
 - {'var2': 'var2_default'} != {'var2': '$encrypted$'}
    - need to look into this, looks like things are not getting hidden, could be a problem with survey_passwords propagation
 - SchemaResourceNotFound: No schema resource '/api/v2/workflow_jobs/166/relaunch/' found. Choices include: ['/api/']
     - completely Latin to me

important tests:

test_wfjt_survey_with_required_and_optional_fields
test_workflow_job_with_project_update
test_workflow_job_template_node_cannot_be_created_without_job_template

AlanCoding committed:
fix bug that broke combining WFJT and node vars

AlanCoding committed:
fix bugs where ask_ var was checked on node

AlanCoding committed:
generalize schedule prompts validation

This makes ScheduleSerializer behave same as WFJT nodes
Prevents providing job_type for workflow jobs, as example
